Sources/AuthFoundation/Token Management/Internal/UserDefaultsTokenStorage.swift

```diff
@@ -24,119 +25,121 @@ class UserDefaultsTokenStorage: TokenStorage {
         }
+        userDefaults.synchronize()
+        delegate?.token(storage: self, defaultChanged: id)
     }
     
-    private(set) var allTokens: [Token] = []
+    var allIDs: [UUID] {
+        Array(allTokens.keys)
+    }
     
-    func contains(token: Token) -> Bool {
-        allTokens.contains(token)
+    func get(token id: UUID) throws -> Token {
+        guard let token = allTokens[id] else {
+            throw TokenError.tokenNotFound(id: id)
+        }
+        
+        return token
     }
     
-    func add(token: Token) throws {
-        guard !allTokens.contains(token) else {
+    func add(token: Token, with id: UUID) throws {
+        guard !allTokens.keys.contains(id) else {
             throw TokenError.duplicateTokenAdded
         } 
 
         var changedDefault = false
         if allTokens.count == 0 {
-            _defaultToken = token
             changedDefault = true
         }
         
-        allTokens.append(token)
+        allTokens[id] = token
         
         try save()
-        delegate?.token(storage: self, added: token)
+        delegate?.token(storage: self, added: id, token: token)
         
         if changedDefault {
-            delegate?.token(storage: self, defaultChanged: token)
+            try setDefaultTokenID(id)
         }
     }
     
-    func replace(token: Token, with newToken: Token) throws {
-        guard let index = allTokens.firstIndex(of: token) else {
+    func replace(token id: UUID, with token: Token) throws {
+        guard let oldToken = allTokens[id] else {
             throw TokenError.cannotReplaceToken
         }
-        
-        allTokens.remove(at: index)
-        allTokens.insert(newToken, at: index)
+        allTokens[id] = token
             
-        var changedDefault = false
-        if _defaultToken == token {
-            changedDefault = true
-            _defaultToken = newToken
+        try save()
+        delegate?.token(storage: self, replaced: id, from: oldToken, to: token)
+    }
+    
+    func remove(id: UUID) throws {
+        guard allTokens[id] != nil else {
+            return
         }
         
-        try save()
-        delegate?.token(storage: self, replaced: token, with: newToken)
+        allTokens.removeValue(forKey: id)
         
-        if changedDefault {
-            delegate?.token(storage: self, defaultChanged: token)
+        try save()
+        delegate?.token(storage: self, removed: id)
+
+        if defaultTokenID == id {
+            try setDefaultTokenID(nil)
         }
     }
     
-    func remove(token: Token) throws {
-        guard let index = allTokens.firstIndex(of: token) else { return }
-        allTokens.remove(at: index)
-        
-        if _defaultToken == token {
-            _defaultToken = nil
-            delegate?.token(storage: self, defaultChanged: nil)
-        }
+    func assign(metadata: [String:String], for id: UUID) throws {
```

Comment:
For me had to read this before I understood what it meant
```suggestion
    func assignMetadata(_ metadata: [String:String], for id: UUID) throws {
```

Reply:
Or, I should go back to my original `setMetadata` name...